frame/treasury/src/lib.rs

```diff
@@ -299,49 +324,80 @@ impl<T: Trait> Module<T> {
-		T::ProposalBondMinimum::get().max(T::ProposalBond::get() * value)
+	fn calculate_bonds(ring: RingBalance<T>, kton: KtonBalance<T>) -> (RingBalance<T>, KtonBalance<T>) {
+		let mut ring_bond: RingBalance<T> = RingBalance::<T>::from(0);
+		let mut kton_bond: KtonBalance<T> = KtonBalance::<T>::from(0);
+
+		if ring > ring_bond {
+			ring_bond = T::RingProposalBondMinimum::get().max(T::ProposalBond::get() * ring);
+		}
+
+		if kton > kton_bond {
+			kton_bond = T::KtonProposalBondMinimum::get().max(T::ProposalBond::get() * kton);
+		}
+
+		(ring_bond, kton_bond)
 	}
 
 	// Spend some money!
 	fn spend_funds() {
-		let mut budget_remaining = Self::pot();
-		Self::deposit_event(RawEvent::Spending(budget_remaining));
+		let mut budget_remaining_ring = Self::pot::<T::RingCurrency>();
+		let mut budget_remaining_kton = Self::pot::<T::KtonCurrency>();
+		let mut imbalance_ring = <RingPositiveImbalance<T>>::zero();
+		let mut imbalance_kton = <KtonPositiveImbalance<T>>::zero();
+		let mut should_burn_ring = true;
+		let mut should_burn_kton = true;
+
+		Self::deposit_event(RawEvent::Spending(budget_remaining_ring, budget_remaining_kton));
 
-		let mut missed_any = false;
-		let mut imbalance = <RingPositiveImbalance<T>>::zero();
 		Approvals::mutate(|v| {
 			v.retain(|&index| {
-				// Should always be true, but shouldn't panic if false or we're screwed.
-				if let Some(p) = Self::proposals(index) {
-					if p.value <= budget_remaining {
-						budget_remaining -= p.value;
-						<Proposals<T>>::remove(index);
-
-						// return their deposit.
-						let _ = T::RingCurrency::unreserve(&p.proposer, p.bond);
-
-						// provide the allocation.
-						imbalance.subsume(T::RingCurrency::deposit_creating(&p.beneficiary, p.value));
-
-						Self::deposit_event(RawEvent::Awarded(index, p.value, p.beneficiary));
-						false
-					} else {
-						missed_any = true;
-						true
+				// Should always be some, but shouldn't panic if false or we're screwed.
+				let mut should_return = false;
+				let option_proposal = Self::proposals(index);
+				if option_proposal.is_none() {
+					return false;
+				}
+
+				let p = option_proposal.unwrap();
+				if p.ring_value > budget_remaining_ring || p.ring_value == RingBalance::<T>::from(0) {
+					should_burn_ring = false;
+					should_return = true;
+				} else {
+					budget_remaining_ring -= p.ring_value;
+					let _ = T::RingCurrency::unreserve(&p.proposer, p.ring_bond);
+					imbalance_ring.subsume(T::RingCurrency::deposit_creating(&p.beneficiary, p.ring_value));
+				}
+
+				if p.kton_value > budget_remaining_kton || p.kton_value == KtonBalance::<T>::from(0) {
+					should_burn_kton = false;
+					if should_return {
+						return true;
 					}
 				} else {
-					false
+					budget_remaining_kton -= p.kton_value;
+					let _ = T::KtonCurrency::unreserve(&p.proposer, p.kton_bond);
+					imbalance_kton.subsume(T::KtonCurrency::deposit_creating(&p.beneficiary, p.kton_value));
```

Comment:
If not missing any, then will enter the burn process.
If budget is not enough, then should not be retained.

For here budget not enough, I think the requirement should be ``` p.ring_value > budget_remaining_ring || p.kton_value > budget_remaining_kton```, so it is different with the logic in the implementation.

Reply:
Fixed.